File: JavaScript/Arrays/AssociativeArray/AssociativeArray.js
/*******************************************************
 *  AssociativeArray.js
 *  Created by Stephen Hall on 11/15/17.
 *  Copyright (c) 2017 Stephen Hall. All rights reserved.
 *  Associative Array implementation in JavaScript
 *  Update Date      Name            Description
 *  1      10/28/19  Stephen Hall    Updated to class syntax
 ********************************************************/

/**
 * AssociativeArray class
 */
class AssociativeArray {
  /**
   * AssociativeArray class Constructor
   * @param {int} size
   * @constructor
   */
  constructor(size) {
    /**
     * Node class for associative array
     * @param key: key to the node
     * @param value: value of the node
     * @param hash: hash of the node
     * @constructor
     */
    this.Node = class {
      constructor(key, value, hash) {
        this.key = key;
        this.value = value;
        this.hash = hash;
      }
    };

    if (size !== undefined && size !== null && size > 0) this.size = size;
    else this.size = 4;
    this.table = [];
    this.count = 0;
  }

  /**
   * Helper method: copies Javas hash code except doesn't allow negative values
   * @param {*} str: string to be hashed
   * @returns {int}: hash of the string
   */
  _hashCode(str) {
    var hash = 0;
    if (str.length === 0) return hash;
    for (var i = 0; i < str.length; i++) {
      var char = str.charCodeAt(i);
      hash = (hash << 5) - hash + char;
      hash = hash & hash; // Convert to 32bit integer
    }
    return Math.abs(hash);
  }

  /**
   * Adds or updates Key, Value pair into the Array
   * @param {*}: Key to associate vale with
   * @param {*}: value to store
   * @returns {Node}: Node added or updated
   */
  set(key, value) {
    var hash = this._hashCode(key.toString());
    var bucket = this._getBucket(hash);
    var entry;
    if (this.isEmpty()) {
      entry = new this.Node(key, value, hash);
      this.table[bucket] = entry;
      this.count++;
    } else {
      entry = this.table[bucket];
      if (!entry) {
        var node = new this.Node(key, value, hash);
        this.table[bucket] = node;
        this.count++;
        entry = node;
      } else {
        while (entry.next !== null) {
          if (this._hashCode(entry) === hash && entry.key === key) {
            entry.value = value;
            return entry;
          }
          entry = entry.next;
        }

        var node = new this.Node(key, value, hash);
        entry.next = node;
        this.count++;
        entry = node;
      }
    }
    return entry;
  }

  /**
   *  Gets the value of the given key
   * @param {*} key
   * @returns {*}: value of the key
   */
  get(key) {
    var hash = this._hashCode(key.toString());
    var bucket = this._getBucket(hash);

    var entry = this.table[bucket];
    while (entry !== null) {
      /* If hash and key matches, return the value */
      if (entry.hash === hash && entry.key === key) {
        return entry.value;
      }
      entry = entry.next;
    }
    return null;
  }

  /**
   * Gets the size of the array
   * @returns {int}
   */
  size() {
    return this.size;
  }

  /**
   * Checks if the array is empty of not
   * @returns {boolean}: true|false
   */
  isEmpty() {
    return this.count === 0;
  }

  /**
   * Gets the bucket index of the hash
   * @param {int} hash
   * @returns {int}
   */
  _getBucket(hash) {
    return hash % this.size;
  }
}

// Tests

let arr = new AssociativeArray(5);
arr.set("taco", "bell");
console.log(arr.get("taco"));
arr.set("king", "kong");
console.log(arr.get("king"));
arr.set("life", 42);
console.log(arr.get("life"));
arr.set(24, 72);
console.log(arr.get(24));

File: Visual_Basic/Arrays/AssociativeArray/AssociativeArray.vb
'******************************************************
' *  AssociativeArray.vb
' *  Created by Stephen Hall on 11/20/17.
' *  Copyright (c) 2017 Stephen Hall. All rights reserved.
' *  Associative Array implementation in vb
' *******************************************************

Namespace Arrays.AssociativeArray
	''' <summary>
	''' Associative Array class declaration
	''' </summary>
	''' <typeparam name="TKey"></typeparam>
	''' <typeparam name="TValue"></typeparam>
	Public Class AssociativeArray(Of TKey, TValue)

		''' <summary>
		''' Node class declaration
		''' </summary>
		Public Class Node
            ''' <summary>
            ''' Public properties of the node class
            ''' </summary>
			Public Property Key As TKey
			Public Property Value As TValue
			Public Property [Next] As Node
			Public Property Hash As Integer

			''' <summary>
			''' Node class Constructor
			''' </summary>
			''' <param name="key1">key of the node</param>
			''' <param name="value2">value for node to hold</param>
			''' <param name="hash3">hash index for index</param>
			Public Sub New(key1 As TKey, value2 As TValue, hash3 As Integer)
				Key = key1
				Value = value2
				Hash = hash3
			End Sub

		End Class

		Private ReadOnly _table As Node()
		Private _size As Integer

		''' <summary>
		''' AssociativeArray class Constructor
		''' </summary>
		Public Sub New()
			_table = New Node(9) {}
			_size = 0
		End Sub

		''' <summary>
		'''  AssociativeArray class Constructo
		''' </summary>
		''' <param name="size">Size to initialize array to</param>
		Public Sub New(size As Integer)
			_table = New Node(size - 1) {}
			_size = 0
		End Sub

		''' <summary>
		''' Adds or updates Key, Value pair into the Array
		''' </summary>
		''' <param name="key">Key to associate vale with</param>
		''' <param name="value">value to store</param>
		''' <returns>Node added or updated</returns>
		Public Function [Set](key As TKey, value As TValue) As Node
			' Find the hash of the key and bucket it belongs to 
			Dim hash As Integer = key.GetHashCode()
			Dim bucket As Integer = GetBucket(hash)
			Dim entry As Node
			If IsEmpty() Then
				entry = New Node(key, value, hash)
				_table(bucket) = entry
				_size += 1
			Else
				entry = _table(bucket)
				While entry.[Next] IsNot Nothing
					If entry.GetHashCode() = hash AndAlso entry.Key.Equals(key) Then
						entry.Value = value
						Return entry
					End If
					entry = entry.[Next]
				End While

				Dim node As New Node(key, value, hash)
				entry.[Next] = node
				_size += 1
				entry = node
			End If
			Return entry
		End Function

		''' <summary>
		''' Gets the value of the given key
		''' </summary>
		''' <param name="key">Key to get value of</param>
		''' <returns>value of the key</returns>
		Public Function [Get](key As TKey) As TValue
			Dim hash As Integer = key.GetHashCode()
			Dim bucket As Integer = GetBucket(hash)

			Dim entry As Node = _table(bucket)
			While entry IsNot Nothing
				' If hash and key matches, return the value 

				If (entry.Hash = hash) AndAlso entry.Key.Equals(key) Then
					Return entry.Value
				End If
				entry = entry.[Next]
			End While
			Return Nothing
		End Function

		''' <summary>
		''' Gets the size of the array
		''' </summary>
		''' <returns>number of elements in the array</returns>
		Public Function Size() As Integer
			Return _size
		End Function

		''' <summary>
		''' Checks if the array is empty of not
		''' </summary>
		''' <returns>true|false</returns>
		Public Function IsEmpty() As Boolean
			Return _size = 0
		End Function

		''' <summary>
		''' Gets the bucket container for the internal array
		''' </summary>
		''' <param name="hash">hash to find bucket of</param>
		''' <returns>bucket index of the array</returns>
		Private Function GetBucket(hash As Integer) As Integer
			Return hash Mod _table.Length
		End Function
	End Class
End Namespace